### PR TITLE
Add support for Waveshare 2.13inch V3 Rev2.1

### LINKED ISSUE #1039: waveshare_3 support?
i just notice my waveshare is a v3
the ink display dont initialise, someone can help please?

### PR DESCRIPTION
**Pull request by ikornaselur**
Adds support for the V3 Rev2.1 of the Waveshare 2.13inch PHAT.

## Description

[Waveshare seems to be phasing out V2 and replacing it with V3](https://www.waveshare.com/wiki/2.13inch_e-Paper_HAT), according to their docs:

> Due to the chip upgrade, the 2.13inch e-Paper V2 will be upgraded to the 2.13inch e-Paper V3 version. The new and old demos are not compatible. Comparing the V3 version with the V2 version, in addition to the chip and driver demo, other appearances and functions are compatible. If you are an old customer and need to purchase a new screen again, please pay attention to the update program. If you are purchasing for the first time, and there is a V3 label on the back of the screen, just use the V3 demo directly. (December 2021)

This PR introduces support for the V3 screen

The library code is provided from https://github.com/waveshare/e-Paper

## Motivation and Context
As V2 is being phased out and V3 is becoming more common, support for it makes sense.
- [ ] I have raised an issue to propose this change ([required](https://github.com/evilsocket/pwnagotchi/blob/master/CONTRIBUTING.md))

Given the simplicity of the change and existance of #1039 I did not raise another issue. Happy to do so for further discussion if requested.

Fixes #1039

## How Has This Been Tested?
Test steps involved:

1. Flash a fresh 1.5.5 on a new SD card with minial config (from the Configuration section on the website)
2. Boot up on Raspberry Pi Zero W with the Waveshare 2.13inch V3 screen connected
3. Set everything up to SSH into the PiZero
4. Shut down the pwnagotchi service
5. Replace the source located in `/usr/local/lib/python3.7/site-packages/pwnagotchi` with the changes from this PR
6. Restart the pwnagotchi service and verify screen is correctly initialised and pwnagotchi displayed on it

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/evilsocket/pwnagotchi/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`

I feel like this support should be documented, but I'm unsure where the documentation change lives. Happy to add info about V3 in the docs if requested!

### CODE REVIEW

**Comment by Langoor2** [comment]
Fixed my issue with the display! Was very confused at first why 213d didn't work and thought the display was dead for a good moment, then I found this issue.

**Comment by viniciuspereiras** [comment]
THANKS, worked for me! xD

**Comment by cryptoads** [comment]
Thank you for this. Still having some issues getting my v3 to work.
Seems that when I add the following to the config, pwnagotchi hangs:

ui.display.type = "waveshare_3"
ui.display.color = "black"

Any thoughts?

**Comment by ChutMei** [comment]
I am very new to this and Have a V3 screen but have no clue how to set this up idk what im doing wrong

**Comment by SeriousM** [comment]
Even I'm sure I ordered the V2, I got the V3 instead after a good week of shipping.
Could we please get this merged so we can use this display officially?

**Comment by viniciuspereiras** [comment]
@ChutMei and @SeriousM, I posted on comments of this thread a steb by step about how to configure a pwnagotchi using waveshare v3 and this pull request :D

**Comment by SeriousM** [comment]
Very nice, thank you @viniciuspereiras. I first tried to build a new image but it seems that the https://github.com/evilsocket/pwnagotchi/pull/1074 would be required as the build didn't succeed. Replacing the python script files is a much easier approach until the new version is (hopefully soon) released.

**Comment by chiotas** [comment]
> **EDIT**
> 
> I decided to download the source and just upload the whole pwnagotchi folder to dist-packages and everything is working now. I must have goofed something up by uploading the individual files.
> 
> Looks good!! thank you.

unfortunately not working for me :(
Sorrym, how can you download the entire pwnagotchi folder?

**Comment by chiotas** [comment]
Thanks, after downloading the whole pwnagotchi folder finally I got an almost working V3 display..However it looks like there is "no ink": the text is extremely dull/washed out..anybody else experienced the same? (Tried with different power sources)

**Comment by chiotas** [comment]
I followed the guide and finally I made the V3 screen work.
Unfortunately it seems hat I get a sort of washed out / faded ink: [https://drive.google.com/file/d/1sa6pE6AJK248xrydgokooqb4762LybVw/view?usp=sharing](url)
Any idea on how to fix?

**Comment by SeriousM** [comment]
Yep, mine looks the same in MANU mode. Start in AUTO mode and the screen update will work.

**Comment by chiotas** [comment]
EPIC, now V3 it's working fine, thanks!

**Comment by HoursAndDreams** [comment]
Hi, can anyone help me to make my V3 work.
I made it to point 3 "Set everything up to SSH into the PiZero"
I work on Windows, pls explaint it to me as easy as possbile.

**Comment by HoursAndDreams** [comment]
I followed the guid and uploaded the new files to the pi, but my display is still not working.
Do I have to do something else after replacing the new files?
Do I have to make a new config.yml file with ui.display.type = "waveshare_3" ?

**Comment by chiotas** [comment]
Didi you replace only 6 files or the whole pwnagotchi folder? Replace the whole folder in case, with just the 6 files listed here under "file changes" is not enough, but with the whole pwnagotchi folder works fine.
Yes you have to set  ui.display.type = "waveshare_3", but you can do it also from the web UI

**Comment by timmosaurusrex** [comment]
Can someone help me? I think i did everything correct and when I look into de log file it stops at this.
`[2022-02-25 12:34:53,816] [INFO] initializing waveshare v3 display`

btw I am using a test image from #1046 on my raspberry pi zero 2 w

**Comment by freesofia** [comment]
pi@pwnagotchi:~ $ sudo /usr/bin/pwnagotchi-launcher
Traceback (most recent call last):
  File "/usr/local/bin/pwnagotchi", line 136, in <module>
    from pwnagotchi.ui.display import Display
  File "/usr/local/lib/python3.7/dist-packages/pwnagotchi/ui/display.py", line 6, in <module>
    import pwnagotchi.ui.hw as hw
  File "/usr/local/lib/python3.7/dist-packages/pwnagotchi/ui/hw/__init__.py", line 9, in <module>
    from pwnagotchi.ui.hw.waveshare3 import WaveshareV3
**ModuleNotFoundError: No module named 'pwnagotchi.ui.hw.waveshare3'**
pi@pwnagotchi:~ $

**Comment by timmosaurusrex** [comment]
Can you please make support for the 2in13b_V3? The b is the colored version.

**Comment by C-Noob404** [comment]
Thank you very much for this solution, but it’s unfortunately almost unusable in Auto Mode, since it’s very washed out. Is there a fix for this?

**Comment by TQMatvey** [comment]
the only real solution for my issu with v3 display, MERGEE

**Review by Katazui — APPROVED** [review]

**Review by azdralovic — APPROVED** [review]

**Review by Allstreamer — APPROVED** [review]

**Comment by TzZek** [comment]
I ordered v2's specifically for my shop, we all received v3's hoping this get's merged so I can simply go over with them how to get this working, from the official repo w/o any extra tinkering from PR's.

**Comment by robot556787** [comment]
hello, I'm very new to this, I'm having a hard time understanding which files to download and where to put them. I know there are steps that will tell you, but I don't understand them at all. Can anyone help me?

**Comment by retzy1989** [comment]
how do you replace the whole pwnagotchi folder? Sorry still new to the Pwnagotchi and would love to see my boy's face!!!

**Comment by TzZek** [comment]
> how do you replace the whole pwnagotchi folder? Sorry still new to the Pwnagotchi and would love to see my boy's face!!!
> 
> 

I mean you could just clone the pwnagotchi folder in the users forked repo but it would be simpler to just follow the above and replace the necessary files. As the project spans many directories in different locations.

**Comment by retzy1989** [comment]
so clone the pwnagotchi folder to /usr/local/lib/python3.7/site-packages/pwnagotchi and then edit the files?

**Comment by retzy1989** [comment]
> dist-packages

Thanks TzZek for the help got it working!!!!

**Comment by rsc-vmartins** [comment]
Good evening guys, I edited all the files as discussed. When I access the logs I see that my pwnagotchi is not getting past the action "initializing waveshare v3 display". How can I solve this problem to get image on waveshare?
image

**Comment by KennyNova** [comment]
i am have the same problem as rsc-vmartins and when i run /usr/bin/pwnagotchi-launcher it says "waveshare_v3 is not supported"

**Comment by KennyNova** [comment]
hey I'm still having problems trying to setup the waveshare v3. 

After mounting the image I put the basic config file but change ui.display.type to wavesharev3.
I kill the pwnagotchi python script and then use file zilla to transfer the 6 files in this pr but then after rebooting the pi it hangs at init wavesharev3 in the logs.

if someone could please help i've been stuck on this for the past few days

**Comment by TzZek** [comment]
> hey I'm still having problems trying to setup the waveshare v3.
> 
> After mounting the image I put the basic config file but change ui.display.type to wavesharev3. I kill the pwnagotchi python script and then use file zilla to transfer the 6 files in this pr but then after rebooting the pi it hangs at init wavesharev3 in the logs.
> 
> if someone could please help i've been stuck on this for the past few days

ui.display.type should be set to be waveshare_3 to match the formatting in pwnagotchi/ui/display.py that was added to the commit 

`    def is_waveshare_v3(self):
        return self._implementation.name == 'waveshare_3'
`

**Comment by KennyNova** [comment]
> ui.display.type should be set to be waveshare_3 to match the formatting in pwnagotchi/ui/display.py that was added to the commit
> 
> ` def is_waveshare_v3(self): return self._implementation.name == 'waveshare_3'`

oops, after looking at my config i do have it set as waveshare_3

**Review by mcduffchannel — APPROVED** [review]

**Review by mcduffchannel — APPROVED** [review]

**Comment by thedroppedpocket** [comment]
resolves #1082

**Comment by pcfreaxx** [comment]
i have edited the file with the changes from the commits. i have uploaded the files. if this is getting in cnflict to any rule, i will delete it.
https://1fichier.com/?rczj8bu26wgb5h6lgbjk
http://ddl.to/d/2ZhTT

**Comment by dennischen0** [comment]
> It seems like this driver makes the pwnagotchi's face a bit smaller, the v2 displays seem to show it being bigger, is that some sort of bug or did you intentionally make the text size smaller for the ASCII face

I noticed this as well. I could not figure out how to fix this.

**Comment by XZCendence** [comment]
> > It seems like this driver makes the pwnagotchi's face a bit smaller, the v2 displays seem to show it being bigger, is that some sort of bug or did you intentionally make the text size smaller for the ASCII face
> 
> I noticed this as well. I could not figure out how to fix this.

I covered this in a reddit guide I made: https://www.reddit.com/r/pwnagotchi/comments/u4q18m/how_to_fix_the_small_face_issue_with_waveshare_v3/ 
Basically, you just change a single font size integer in the waveshare3.py layout file in /usr/local/lib/python3.7/dist-packages/pwnagotchi/ui/hw

**Comment by lcrince** [comment]
Still no merge?

**Comment by pcfreaxx** [comment]
Nope. But i have a ready-to-use Image with waveshare v3 support dumped and uploaded to archive.org
https://archive.org/details/pwnagotchi_1.5.5_WSV3Patched

**Comment by unknown-entity-prog** [comment]
@pcfreaxx just out of interest, what would be the contents of the config file? Still a little new to this...

**Comment by pcfreaxx** [comment]
> @pcfreaxx just out of interest, what would be the contents of the config file be? Still a little new to this...

there is no magic.
have a look here:
`main.name = "pwnagotchi"
main.lang = "en"
main.whitelist = [
  "YourWIFI",
]

main.plugins.grid.enabled = true
main.plugins.grid.report = true
main.plugins.grid.exclude = [
  "YourWIFI"
]

ui.display.enabled = true
ui.display.type = "waveshare_3"
ui.display.color = "black"
`

**Comment by unknown-entity-prog** [comment]
> Nope. But i have a ready-to-use Image with waveshare v3 support dumped and uploaded to archive.org
> https://archive.org/details/pwnagotchi_1.5.5_WSV3Patched

This patch works like a dream, make sure you have a cable that supports data... i bought one that stated data transfer but after a few failed attempts i tried another cable that worked!!

**Comment by pcfreaxx** [comment]
You are allowed to use my Image. I am not the Hero, i just worked with the commits from pr

**Comment by unknown-entity-prog** [comment]
I'm not sure whether i have an isolated issue but i am able to SSH into the Pwnagotchi while it is connected to the USB port with the data cable inserted, I am unable to connect wirelessly via SSH as the wlan0 disappears every time after reboot. Is this normal behavior and if so is there a way i am able to change it? I have populated a wpa_supplicant file and loaded it into boot prior to first boot into pwnagotchi but still nothing. Thanks in advance

File: pwnagotchi/ui/hw/waveshare3.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from pwnagotchi.ui.hw.base import DisplayImpl
+
+
+class WaveshareV3(DisplayImpl):
+    def __init__(self, config):
+        super(WaveshareV3, self).__init__(config, 'waveshare_3')
+        self._display = None
+
+    def layout(self):
+        fonts.setup(10, 8, 10, 25, 25, 9)
```
Comment (voidlily) [inline pwnagotchi/ui/hw/waveshare3.py:13]:
```suggestion
        fonts.setup(10, 8, 10, 35, 25, 9)
```

small face fix from https://old.reddit.com/r/pwnagotchi/comments/u4q18m/how_to_fix_the_small_face_issue_with_waveshare_v3/

**Comment by AABatteries** [comment]
> Nope. But i have a ready-to-use Image with waveshare v3 support dumped and uploaded to archive.org https://archive.org/details/pwnagotchi_1.5.5_WSV3Patched

Will this image work with RPi Zero 2W? or is it just for the original RPi Zero W?

**Comment by pcfreaxx** [comment]
I dont own a rpi z2 so i cant say. But i think it is possible when you Just use the pwnagotchi folder. At this Point works python

**Comment by abonforti** [comment]
Does anyone have experienced this problem? Waveshare 2.13 v3.

IMG_5586

**Comment by invictusani** [comment]
@pcfreaxx Thank you for the wavesharev3 file, it worked perfectly! I'm having issues SSHing into the pwnagotchi though. I've tried 'raspberry' in multiple iterations and it's not working. I'm wondering if you know what the default password would be?

**Comment by pcfreaxx** [comment]
@invictusani the default password is raspberry, you are right. I would check config.toml maybe it is still an ip error.

**Comment by invictusani** [comment]
@pcfreaxx sorry but im very fresh to all of this. How do I access the config.toml file? I'm on a Mac and whenever I try to access the boot drive it doesn't show me a config.toml file, but there is a config.txt file. it only contains info for connecting the device via hdmi. I've managed to configure the ip through my network manager on my Mac.

**Comment by pcfreaxx** [comment]
@invictusani i usually work with Linux and Windows. But after some Research i 've found Paragon ExtFS. It should let you enter your rootfs. Normally i mount the Partition on Linux but Paragon offers the same Software for Windows to. Dont touch config.txt, If not need. Your solution is for sure to edit config.toml.

**Comment by invictusani** [comment]
@pcfreaxx I got it to work! Thank you so much for your time and help, it's really appreciated!

**Comment by pcfreaxx** [comment]
@invictusani glad to help, Hack the Planet.

**Comment by Spessmen14** [comment]
> pcfreaxx

I've tried flashing that image with two separate utilitiesand my Pi 0 gets stuck on "Update UMTP about System Runlevel Changes".

How can this problem be alleviated with the patched image?

**Comment by pcfreaxx** [comment]
@Spessmen14 sry you are the first wir that Problem, i have no clue whats going wrong.

**Comment by Spessmen14** [comment]
> @Spessmen14 sry you are the first with that Problem, i have no clue whats going wrong. Can you explain your steps?

1. Flash the image from archive.org to a microSD card, that previously worked with a vanilla Pwn image.
2. Try to boot
3. Get stuck at boot with this message.

**Comment by Spessmen14** [comment]
> @Spessmen14 sry you are the first with that Problem, i have no clue whats going wrong. Can you explain your steps?

I left it alone by itself and EVENTUALLY, like after an hour, it seems to be working properly. In german. But now when connecting a microUSB to the data port it won't let me log in, it doesn't accept any keyboard input. I should perhaps mention this is connected to a HAT.

**Comment by XZCendence** [comment]
I covered this in my reddit guide:
https://www.reddit.com/r/pwnagotchi/comments/u4q18m/how_to_fix_the_small_face_issue_with_waveshare_v3/

**Comment by HowToInternet** [comment]
> Nope. But i have a ready-to-use Image with waveshare v3 support dumped and uploaded to archive.org https://archive.org/details/pwnagotchi_1.5.5_WSV3Patched

This file isn't able to be opened by winrar on multiple machines through either the direct .rar download or the torrent version. Winrar advises the following: "The archive is either in unknown format or damaged"

**Comment by HowToInternet** [comment]
> Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.

It worked after downloading just the image in the contents of the rar file. Strange. Now I'm unable to SSH into my pi after flashing this image due to the host key being changed. Are you familiar with how I'd get around this? I'm just using the SSH command in PowerShell, not using Putty or another client.

**Comment by HowToInternet** [comment]
> Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.

No worries at all. I was able to just download the .img instead of the entire .rar library and that seemed to get things up and running. I was able to SSH in and then change the language and my home network settings. Its name hasn't update for some reason but I'll figure that out later. I think the last thing I just need to figure out is how to fix the issue with the screen looking like its faded or dim.

**Comment by Crash891** [comment]
> Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.

hello, i tried with your image, to use 'raspberry' as password in SSH connection, but it don't let me enter. i can't enter browsing 10.0.0.2 too, cause it want username and password. any hints?

**Comment by timmosaurusrex** [comment]
> > Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.
> 
> 
> 
> hello, i tried with your image, to use 'raspberry' as password in SSH connection, but it don't let me enter. i can't enter browsing 10.0.0.2 too, cause it want username and password. any hints?
> 
> ps: i managed to do with the normal image of pwnagotchi, but obv the screen didn't worked. now i can see the PING, but it refuse to let me in

Do you use pi as username? I use pi@10.0.0.2 with password "raspberry"

**Comment by Crash891** [comment]
> > > Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.
> > 
> > 
> > hello, i tried with your image, to use 'raspberry' as password in SSH connection, but it don't let me enter. i can't enter browsing 10.0.0.2 too, cause it want username and password. any hints?
> > ps: i managed to do with the normal image of pwnagotchi, but obv the screen didn't worked. now i can see the PING, but it refuse to let me in
> 
> Do you use pi as username? I use pi@10.0.0.2 with password "raspberry"

oh yes, i solved it using pi@10.0.0.2, psw raspberry. now there are other problems... for example i can't do "sudo pip3 install --upgrade pwnagotchi". I'm in V1.5.3, cause in 1.5.5 vanilla the pwnagotchi have some issue with AI, so (like some suggestion in other GitHub issue) i want to upgrade it from the V1.5.3, but it give me "Requirement already satisfied" or other connection problems like "Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0xb5566650>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/pwnagotchi/"   (it is connected via USB, and it can ping 1.1.1.1, so it should be connected to internet)

**Comment by Crash891** [comment]
> > > Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.
> > 
> > 
> > hello, i tried with your image, to use 'raspberry' as password in SSH connection, but it don't let me enter. i can't enter browsing 10.0.0.2 too, cause it want username and password. any hints?
> > ps: i managed to do with the normal image of pwnagotchi, but obv the screen didn't worked. now i can see the PING, but it refuse to let me in
> 
> Do you use pi as username? I use pi@10.0.0.2 with password "raspberry"

another problem is that pwnagotchi random reboot giving
----------- 
"Task exception was never retrieved
future: <Task finished coro=<Client.start_websocket() done, defined at /usr/local/lib/python3.7/dist-packages/pwnagotchi/bettercap.py:38> exception=OSError("Multiple exceptions: [Errno 111] Connect call failed ('::1', 8081...
Traceback (most recent call last):
File "/usr/local/lib/python3.7/dist-packages/pwnagotchi/bettercap.py", line 42, in start_websocket
async with websockets.connect(s, ping_interval=60, ping_timeout=90) as ws:
File "/usr/local/lib/python3.7/dist-packages/websockets/client.py", line 517, in __aenter__
return await self
File "/usr/local/lib/python3.7/dist-packages/websockets/client.py", line 535, in __await_impl__
transport, protocol = await self._create_connection()
File "/usr/lib/python3.7/asyncio/base_events.py", line 968, in create_connection
', '.join(str(exc) for exc in exceptions)))
OSError: Multiple exceptions: [Errno 111] Connect call failed ('::1', 8081, 0, 0), [Errno 111] Connect call failed ('127.0.0.1', 8081)"
----------

**Comment by Crash891** [comment]
> > > Sorry but much, much people load it and it didnt changed since first upload. I tried it a few minutes ago and there we're no Problems.
> > 
> > 
> > hello, i tried with your image, to use 'raspberry' as password in SSH connection, but it don't let me enter. i can't enter browsing 10.0.0.2 too, cause it want username and password. any hints?
> > ps: i managed to do with the normal image of pwnagotchi, but obv the screen didn't worked. now i can see the PING, but it refuse to let me in
> 
> Do you use pi as username? I use pi@10.0.0.2 with password "raspberry"

another problem is that it reboot random, giving in logtails:
"Task exception was never retrieved
future: <Task finished coro=<Client.start_websocket() done, defined at /usr/local/lib/python3.7/dist-packages/pwnagotchi/bettercap.py:38> exception=OSError("Multiple exceptions: [Errno 111] Connect call failed ('::1', 8081...
Traceback (most recent call last):
File "/usr/local/lib/python3.7/dist-packages/pwnagotchi/bettercap.py", line 42, in start_websocket
async with websockets.connect(s, ping_interval=60, ping_timeout=90) as ws:
File "/usr/local/lib/python3.7/dist-packages/websockets/client.py", line 517, in aenter
return await self
File "/usr/local/lib/python3.7/dist-packages/websockets/client.py", line 535, in await_impl
transport, protocol = await self._create_connection()
File "/usr/lib/python3.7/asyncio/base_events.py", line 968, in create_connection
', '.join(str(exc) for exc in exceptions)))
OSError: Multiple exceptions: [Errno 111] Connect call failed ('::1', 8081, 0, 0), [Errno 111] Connect call failed ('127.0.0.1', 8081)"

**Comment by Sirius-RJ** [comment]
> With some people on the unofficial discord, we decided to start a new fork with all new fixes in. Do you want to join us ? https://github.com/pwnagotchi2/pwnagotchi

page not found !

**Comment by Paranoi8972** [comment]
I'm really happy it worked. But is there a way to add costume plugins? I know, I can edit the standard plugins in the webconfig `"http://10.0.0.2:8080/plugins/webcfg`. If I do it in the `config.toml` the screen stops working but this is, so far I know, required to add the path for the plugins. 
Thanks!

**Comment by azdralovic** [comment]
> 

You should do it in config.toml, but not sure why your screen stops working... works fine for me

**Comment by Paranoi8972** [comment]
Nah, this was me too frustrating. It all just won’t work for me. I just bought a v2. So I’m good with everything. Everything works. I’m happy. 
Thanks

**Comment by Miente1** [comment]
I changed the files to support my V3 display. But when i reboot the pwnagotchi it does not work.
Also web ui an BT-tethering stops working.

Only way to get it back to work is put the display enable back to false.

Any work around on this?

**Comment by niyazaki** [comment]
As this is Merged, setting the screen to "waveshare_3" should do the trick right ?
It doesn't, at least for me. Anyone having the same issue ?

I tried waveshare213d as I saw on Reddit, but not working either :(.

Can someone help ? Thank you !

**Comment by lorenzoPrimi** [comment]
Same problem here, I tried to install the 1.55 and clone the current master and also tried the `pwnagotchi_1.5.5_WSV3Patched`
Nothing is showing up, this is the log :

```
[2019-07-10 01:21:43,627] [INFO] [update] plugin loaded.
[2019-07-10 01:21:43,630] [INFO] [led] plugin loaded for /sys/class/leds/led0/brightness
[2019-07-10 01:21:43,635] [INFO] grid plugin loaded.
[2019-07-10 01:21:43,633] [INFO] Session-stats plugin loaded.
[2019-07-10 01:21:43,773] [WARNING] ui.fps is 0, the display will only update for major changes
[2019-07-10 01:21:43,778] [INFO] initializing waveshare v3 display
[2019-07-10 01:21:47,335] [INFO] generating /etc/pwnagotchi/id_rsa ...
```
Nothing is displayed. 
Could my display be broken?

**Comment by uzonder** [comment]
> Thank you for this. Still having some issues getting my v3 to work. Seems that when I add the following to the config, pwnagotchi hangs:
> 
> ui.display.type = "waveshare_3" ui.display.color = "black"
> 
> Any thoughts?
> 
> **EDIT**
> 
> I decided to download the source and just upload the whole pwnagotchi folder to dist-packages and everything is working now. I must have goofed something up by uploading the individual files.
> 
> Looks good!! thank you.

Replacing the whole pwnagotchi folder from the source resolved the issue for me. Individual files did not work.

**Comment by yb278** [comment]
Hello, I have never used ssh before and dont get how to actually replace the whole folder in terms of what commands to use. I can ssh into it and browse directories but could someone tell me what to actually do to change the whole folder

**Comment by timmosaurusrex** [comment]
> Hello, I have never used ssh before and dont get how to actually replace the whole folder in terms of what commands to use. I can ssh into it and browse directories but could someone tell me what to actually do to change the whole folder

I use WinSCP to copy files to my raspberry. First i remove the directory with "sudo rm -rf" (this will permanently remove the directory and everything inside). Then I drag and drop the files i want to copy in WinSCP.

**Comment by Tr33x0rs** [comment]
I recently learned about this project and decided to build one of my own. I initially ordered a waveshare 2.13 e ink display and had issues getting it to work but finally got it to work with the evil socket branch. Unfortunately I damaged the displays ribbon cable when trying to insert it into a case and it no longer works.

I purchased another display but mistakenly it is the Touch version but is also a V3. My config file is still the same config file that I was using before which is showing waveshare 3 as shown on the evil socket GitHub. 

Pwnagotchi version 1.55 by evilsocket- https://github.com/evilsocket/pwnagotchi/releases

my config.toml -

main.name = "pwnagotchi" main.lang = "en" main.whitelist = [ "EXAMPLE_NETWORK", "ANOTHER_EXAMPLE_NETWORK", "fo:od:ba:be:fo:od", "fo:od:ba" ]

main.plugins.grid.enabled = true main.plugins.grid.report = true main.plugins.grid.exclude = [ "YourHomeNetworkHere" ]

ui.display.enabled = true ui.display.type = "waveshare_3" ui.display.color = "black"

I followed the instructions of downloading the pwnagotchi-master files and replacing the 6 files in their appropriate directories (utils.py to /pwnagotchi/, waveshare3.py to /pwnagotchi/ui/hw/   etc..). I waited about 20 mins and restarted my pi. I'm still getting nothing on the display. 

I am able to ssh into the pi no problem. Please help. I've been tinkering with this for days and it has been driving me mad! Thank you in advance.

**Comment by 2lag** [comment]
@Tr33x0rs did you ever solve this? i'm having the same issue and have been for too long now

**Comment by costyn** [comment]
For anyone still struggling, check out the aluminum-ice fork of this project. There you can set the config to waveshare_v3 and it should work... give it a few minutes after boot before the screen turns on.

**Pull request merged**